Firestore/core/src/firebase/firestore/util/status.h

```diff
@@ -128,9 +144,49 @@ class ABSL_MUST_USE_RESULT Status {
     // NSError* to Status and back to NSError* losslessly.
     std::unique_ptr<PlatformError> platform_error;
   };
-  // OK status has a `nullptr` state_.  Otherwise, `state_` points to
-  // a `State` structure containing the error code and message(s)
-  std::unique_ptr<State> state_;
+
+  // A `unique_ptr` with a custom deleter. If the pointer's value has been set
+  // to a special value (0x1) to indicate it is moved, the custom deleter will
+  // stop `unique_ptr` trying to destrcut from an invalid address.
+  using StatePtr = std::unique_ptr<State, std::function<void(State*)>>;
```

Comment:
Question: now that you've added a constructor taking two parameters to `State`, it seems like `State` can be immutable? Can you try making this `std::unique_ptr<const State, Deleter>`? (i.e., a pointer to const `State`). If it turns out to be hard to do, just ignore.

Reply:
There are couple of assignment to `state_->platform_error` prevent this.